### ISSUE TITLE
CPS File Progress Report

### ISSUE THREAD

**Issue by andersonfrailey** [open-source-economics/taxdata#90]
This issue is just an overview of the progress we've made in preparing the CPS-based file for use in Tax-Calculator.

John gave me the files needed to create the CPS file along with an associated weights file that covers the years 2015-2027. The SAS scripts create tax-units from the CPS in the same manner used to create the CPS tax-units that are then merged with the 2009 IRS-PUF file to create the final PUF currently used. After that, the following files are adjusted for top-coding:

* Wages and salaries
* Taxable interest income
* Dividends
* Alimony
* Business income/loss
* Pensions
* Rents
* Farm income/loss

Then the following are imputed:

* Capital gains
* Taxable IRA distributions
* Adjusted IRA Contributions
* KEOGH/SEP plan contributions
* Self-employed health insurance deduction
* Student loan interest deduction
* Charitable contributions
* Miscellaneous deductions
* Child care expenses
* Medical expenses deduction
* Home mortgage interest expense
* Real estate taxes
* Domestic production activity deduction

Finally, the following are targeted at a state level:
* Wages
* Interest income
* Dividends
* Business income/loss
* Capital gains
* Taxable IRA distributions
* Pensions
* Unemployment
* KEOGH
* Self-employment health insurance
* IRA contribution
* Student loan interest
* Domestic production activity deduction
* Schedule E income

There are some variables that are currently missing from the file as well:

| Variable     | Description                                                         | 
|--------------|---------------------------------------------------------------------| 
| p23250       |  Sch D: Net long-term capital gains/losses                          | 
| p25470       |  Sch E: Royalty depletion and/or rental depreciation                | 
| e09800       |  Unreported payroll taxes from Form 4137 or 8919                    | 
| e02000       |  Sch E rental, royalty, S-corp, etc, income/loss                    | 
| e62900       |  Alternative Minimum Tax foreign tax credit from Form 6251          | 
| p08000       |  Other tax credits (but not including Sch R credit)                 | 
| e58990       |  Investment income elected amount from Form 4952                    | 
| e00700       |  Taxable refunds of state and local income taxes                    | 
| e03290       |  Health savings account deduction from Form 8889                    | 
| e07240       |  Retirement savings contributions credit from Form 8880             | 
| agi_bin      |  Historical AGI category used in data extrapolation                 | 
| e19200       |  Sch A: Interest paid                                               | 
| e27200       |  Sch E: Farm rent net income or loss                                | 
| e01200       |  Other net gain/loss from Form 4797                                 | 
| e03500       |  Alimony paid                                                       | 
| n1821        |  Number of people over 18 and under 21 years old in the filing unit | 
| e07260       |  Residential energy credit from Form 5695                           | 
| blind_spouse |  1 if spouse is blind; otherwise 0                                  | 
| p22250       |  Sch D: Net short-term capital gains/losses                         | 
| e03220       |  Educator expenses                                                  | 
| e07400       |  General business credit from Form 3800                             | 
| f2441        |  number of child/dependent-care qualifying persons                  | 
| nu18         |  Number of people under 18 years old in the filing unit             | 
| f6251        |  1 if Form 6251 (AMT) attached to return; otherwise 0               | 
| blind_head   |  1 if taxpayer is blind; otherwise 0                                | 
| e03230       |  Tuition and fees from Form 8917                                    | 
| e03400       |  Penalty on early withdrawal of savings                             | 
| e07300       |  Foreign tax credit from Form 1116                                  | 
| e11200       |  Excess payroll (FICA/RRTA) tax withheld                            | 
| e24518       |  Sch D: 28% Rate Gain or Loss                                       | 
| EIC          |  number of EIC qualifying children (range: 0 to 3)                  | 
| e09700       |  Recapture of Investment Credit                                     | 
| p87521       |  Total tentative AmOppCredit amount for all students                | 
| e09900       |  Penalty tax on qualified retirement plans                          | 
| e24515       |  Sch D: Un-Recaptured Section 1250 Gain                             | 
| e01500       |  Pensions and annuities                                             | 
| e87530       |  Adjusted qualified lifetime learning expenses for all students     | 
| e26270       |  Sch E: Combined partnership and S-corporation net income/loss      | 
| e07600       |  Prior year minimum tax credit from Form 8801                       | 
| cmbtp        |  Estimate of income on (AMT) Form 6251 but not in AGI               | 
| n21          |  Number of people 21 years old or older in the filing unit          | 
| e18400       |  Sch A: State and local income/sales taxes                          | 
| e20500       |  Sch A: Net casualty or theft loss                                  | 
| MIDR         |  1 if separately filing spouse itemizes; otherwise 0                | 

`nu18, n1821, and n21` can be found, and I'm editing the SAS files to do so. We're waiting for John to get us imputations for state and local taxes as well. I'm also digging more into the CPS to see if there are any other variables that can be found.

@Amy-Xu and I are analyzing the final files to make sure the results after using it in tax-calc makes sense.

I will use this issue to post updates as more progress is made

@martinholmer @MattHJensen @codykallen

**Comment by martinholmer** [comment]
@andersonfrailey said in taxdata issue #90 wrt the being-developed CPS data file:
> [The currently missing variables] `nu18`, `n1821`, and `n21` can be found, and I'm editing the SAS files to do so.

Thanks for the progress report!  Sounds like things are progressing nicely.

I always thought one of the big advantages of CPS data is that you know everybody's age, right?
If so, then why can't you make a pretty good estimate of the value of the following variables for each filing unit?
```
EIC
f2441
```
Without these variables (especially without the EIC variable) the tax estimates generated by this CPS file are going to be way off.

@MattHJensen @Amy-Xu

**Comment by andersonfrailey** [comment]
@martinholmer, you are correct that a big advantage of the CPS is knowing everybody's age. The two variables you listed first, `EIC, f2441` are two that I'll be turning my attention to next. 

And we do have the values for the second set listed.

**Comment by martinholmer** [comment]
@andersonfrailey said:
> you are correct that a big advantage of the CPS is knowing everybody's age. The two variables you listed first, EIC, f2441 are two that I'll be turning my attention to next.
>
> And we do have the values for the second set listed.

Great!

**Comment by martinholmer** [comment]
For earlier discussion (from 03-Nov-2016 to 02-Dec-2016) of issues related to the development of this CPS input file for Tax-Calculator, see [Tax-Calculator issue 1030](https://github.com/open-source-economics/Tax-Calculator/issues/1030).

**Comment by andersonfrailey** [comment]
Quick update on the CPS file. [Here](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb) is a notebook where you can see some comparisons between the CPS and PUF file after running Tax-Calc with each for 2017.

Some notable differences between the two can be seen in regular tax and income tax liability. That could be partially due to a couple of missing income items such as investment income. 

You can also see that itemized deductions are significantly lower in the CPS. As noted before, John is imputing state and local tax deductions so once we have that the difference should close some, but we will still be missing a few deductions.

I also want to add farm income (e02100) and capital gain distributions (e01100) to the missing variables list. After looking at the aggregates for each, they are significantly higher in the CPS than the PUF. This may simply be due to a mislabeling when prepping the CPS. I'm looking into it further. That being said, I am a little more concerned about the drop in tax liabilities given that these two income sources were so much higher than they should have been and yet liabilities still fall.

If there is anything specific you would like to see in my next update, please let me know.

@martinholmer @MattHJensen @Amy-Xu

**Comment by andersonfrailey** [comment]
I've added a few distribution plots for the CPS data to the [notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb). It looks to me like significant stage 3 adjustment will be needed.

**Comment by andersonfrailey** [comment]
John sent me a preliminary version of state and local tax deductions. The [notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb) has been updated to include a chart comparing totals for a number of itemized deductions.

**Comment by Amy-Xu** [comment]
@andersonfrailey Just to confirm all blow-up factors used for CPS-Tax-Unit file is the same ones used for PUF, is that right? Same for the stage-I factors used to calculate weights?

If that's the case, I think we might need to develop a separate stage-I factor, or adjust the base-year file before feeding it to TC due to the large gap for several income/expense items.

**Comment by martinholmer** [comment]
@Amy-Xu asked:
> Just to confirm all blow-up factors used for CPS-Tax-Unit file is the same ones used for PUF, is that right?  Same for the stage-I factors used to calculate weights?

Unless I'm confused, "blow-up factors" and "stage-I factors" are two different names for the same thing.  And yes, they affect the baseline (CBO-derived) projection, so they **must** be the same for the all micro datasets, by definition.

@Amy-Xu continued:
> If that's the case, I think we might need to develop separate stage-I factors [for the CPS and the PUF datasets].

This is not possible given the logic above.

The stage2 weights can differ for the CPS and PUF datasets and the stage3 adjustment factor can be different for CPS and PUF, but not stage1 factors (otherwise the two datasets would be using different CBO baseline projections, which doesn't make any sense).

**Comment by Amy-Xu** [comment]
@martinholmer said
> Unless I'm confused, "blow-up factors" and "stage-I factors" are two different names for the same thing. 

Right I'm aware of that so I said 'just to confirm'.

>And yes, they affect the baseline (CBO-derived) projection, so they must be the same for all micro datasets, by definition.

For this part, it's true that we derive most factors to match CBO baseline, but not every factor is derived from CBO projection. Instead, some of the factors are tuned to fit PUF data. For example, we originally assumed to apply personal income factor (ATXPY) to most PUF variables outside AGI, which is a very reasonable assumption. But then we realized that applying this factor to ATXPY is not proper for e19200, which is interest paid that used to calculate itemized deduction. It's not proper since the growth rate of its major component -- home mortgage interest deduction -- deviates from the personal income growth rate. So we added one extra factor (AIPD) for this variable only to blow up this factor at the rate given by SOI tables prior to 2014, and then apply ATXPY after 2014. This particular factor yields better itemized deduction, which is not predicted by CBO explicitly, and therefore makes our results closer to CBO projected baseline.

When it comes to CPS data, as you can see in Anderson's notebook, home mortgage interest differs from PUF total by about 200 billion (total at ~350 billion in PUF), which means if we use exact same factor (originally derive from SOI tables) for CPS, we will significantly over-estimate all interest paid deduction, and therefore overestimate total number of itemizers, which would drag our CPS-tax-unit results away from CBO baseline. 

Yes I'm aware that having two sets of factors merely to adjust itemize deduction is too much work and may or may not be efficient for the goal (adjusting the baseline to resemble CBO projection) we want to achieve. So alternatively I proposed,

>or adjust the base-year file before feeding it to TC

which I consider as a more doable way for adjusting itemized deduction and related expense projection. Does it make some sense?

**Comment by andersonfrailey** [comment]
@Amy-Xu 
> or adjust the base-year file before feeding it to TC

As in perform an adjustment similar to what we do in stage 3? Or did you have something else in mind?

**Comment by Amy-Xu** [comment]
@andersonfrailey I was thinking to tweak stage I factors a bit like AIPD, but perfectly fine with a Stage III as well. The goal is to take a look whether adjusting the itemize deductions would fix itemizer/standard deductor numbers, and furthur help AMT numbers.

**Comment by martinholmer** [comment]
@Amy-Xu said in issue #90 wrt the CPS dataset:
> For this part, it's true that we derive most factors to match CBO baseline, but not every factor is derived from CBO projection. Instead, some of the factors are tuned to fit PUF data. For example, we originally assumed to apply personal income factor (ATXPY) to most PUF variables outside AGI, which is a very reasonable assumption. But then we realized that applying this factor to ATXPY is not proper for e19200, which is interest paid that used to calculate itemized deduction. It's not proper since the growth rate of its major component -- home mortgage interest deduction -- deviates from the personal income growth rate. So we added one extra factor (AIPD) for this variable only to blow up this factor at the rate given by SOI tables prior to 2014, and then apply ATXPY after 2014. This particular factor yields better itemized deduction, which is not predicted by CBO explicitly, and therefore makes our results closer to CBO projected baseline.
>
> When it comes to CPS data, as you can see in Anderson's notebook, home mortgage interest differs from PUF total by about 200 billion (total at ~350 billion in PUF), which means if we use exact same factor (originally derive from SOI tables) for CPS, we will significantly over-estimate all interest paid deduction, and therefore overestimate total number of itemizers, which would drag our CPS-tax-unit results away from CBO baseline. Same would apply to state and local deduction.
>
> Yes I'm aware that having two sets of factors merely to adjust itemize deduction is too much work and may or may not be efficient for the goal (adjusting the baseline to resemble CBO projection) we want to achieve. So alternatively I proposed,
>
>> or adjust the base-year file before feeding it to TC
>
> which I consider as a more doable way for adjusting itemized deduction and related expense projection. Does it make some sense?

Yes, it is true that not all of the stage1 growfactors are taken from an CBO projection.  The story you tell about why the mortgage interest needed its own growfactor (because of the sharp drop in mortgage interest rates following the 2008-2009 financial crash and house-price deflation) is true.  But that growfactor reflects the reality of the macroeconomic situation in the years after 2009.  And therefore, it is not a logical choice for changing across micro datasets.

And we seem to be in agreement on that point because you say:
> So alternatively I proposed,
>
>> or adjust the base-year file before feeding it to TC
>
> which I consider as a more doable way for adjusting itemized deduction and related expense projection.

Perhaps @andersonfrailey could do a one-time adjustment (to close the gap) at the end of the python code that creates the raw CPS dataset.  And then we can see if the gap remains closed in all projection years.  If the gap becomes too large, @andersonfrailey could use the stage3 adjustment methodology he created to manage the gap in subsequent years.  Or can stage3 methods be used to close the gap in *all* years?  I can't remember whether stage3 can deal with the first year.

Does this seem like a sensible way to proceed in the effort to make the CPS dataset generate results that a reasonably close to those generated by the `puf.csv` dataset?

**Comment by andersonfrailey** [comment]
@Amy-Xu @martinholmer, Stage 3 adjustments wouldn't be the best way to go about this because all stage 3 does is adjust the _distribution_ of the variable, not the total amount.

In my opinion, a one time adjustment like @martinholmer proposed would work best. To be clear, this CPS file is being produced by SAS files provided by John, but this adjustment could be added at the end of the CPS version of `finalprep.py`

**Comment by martinholmer** [comment]
@andersonfrailey said in issue #90 wrt the new CPS-only input dataset:
> Stage 3 adjustments wouldn't be the best way to go about this because all stage 3 does is adjust the distribution of the variable [across AGI bins], not the total amount.
>
> In my opinion, a one time adjustment like @martinholmer proposed would work best. To be clear, this CPS file is being produced by SAS files provided by John, but this adjustment could be added at the end of the CPS version of finalprep.py

Sounds sensible to me.  What do you think, @Amy-Xu ?

**Comment by Amy-Xu** [comment]
@andersonfrailey  said:
>In my opinion, a one time adjustment like @martinholmer proposed would work best. To be clear, this CPS file is being produced by SAS files provided by John, but this adjustment could be added at the end of the CPS version of finalprep.py

@martinholmer followed with:
>Sounds sensible to me. What do you think, @Amy-Xu ?

Yes that sounds good to me!

**Comment by Amy-Xu** [comment]
@andersonfrailey I personally feel it would be valuable to ask John whether this level (we're talking about ~150b difference for both state & local and home mortgage) of discrepancy is expected while implementing this adjustment.

**Comment by andersonfrailey** [comment]
@Amy-Xu I agree. I'll shoot him an email.

**Comment by andersonfrailey** [comment]
In response to my email John pointed out that he did the imputations for _every_  tax unit, where the matched file would only have values for the records that had filed so I was comparing two things that weren't really the same. I've updated the [notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb) to account for this and the state and local deduction looks much better.

**Comment by andersonfrailey** [comment]
As mentioned previously, many of the income variables in the CPS-PUF will need substantial adjustments to get their distribution right. Unlike the PUF file, the CPS-PUF contains records from the 2013, 2014, and 2015 CPS, which correspond to 2012, 2013, and 2014 tax years. The 2013 and 2014 files are aged to match the 2015 file (2014 tax year).

Because so many of the variables need adjusting and they're already equivalent to the 2014 tax year, I'm considering adding a one time adjustment to the `finalprep.py` file. My reasoning is, in stage 3 we assume no change in the distribution of adjusted variables after 2014 and the start year for the CPS file will need to be set at 2014. Therefore, in my opinion, we might as well adjust the variables in the prep stages rather than have to modify tax-calc to account for all of the additional variables that will need adjusting.

Thoughts?

@martinholmer @Amy-Xu @MattHJensen

**Comment by Amy-Xu** [comment]
By adjustment, do you mean augment or shrink the total to match PUF? How many variables do we need adjustment? @andersonfrailey

**Comment by andersonfrailey** [comment]
@Amy-Xu, neither. In this case I'm just talking about fixing the distribution of the variables. Specifically I'd say interest income and ordinary and qualified dividends definitely need adjustments and business income could benefit as well.

The actual totals for the income variables are actually relatively close to the PUF already.

**Comment by Amy-Xu** [comment]
I see, that sounds valuable. At the same time, it might also be helpful to investigate what is the major source dragging individual income tax liability off by almost 40%. The distribution of those income variables could be a reason, but not sure whether they're big problems given totals are close.

**Comment by andersonfrailey** [comment]
I spent the day working on adjusting the distribution of interest and business income and ordinary and qualified dividends. I don't believe it's possible to get the distribution as accurate as we were able to with the PUF, but there is improvement.

AGI is reported in the CPS, but I quickly realized that it would not be very helpful. The highest reported AGI was just over $2 million, which would put nobody in the top two income bins, as defined in the IRS SOI data mentioned previously in this thread. I instead added up wages, interest income, dividends, alimony, business income, pensions, rental income, farm income, unemployment compensation, and social security income.

The new distributions can be seen in the updated [notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb). There was a slight improvement in individual income tax liability. It is now short by about 30% rather than 40%

**Comment by Amy-Xu** [comment]
The number of itemizers and standard deductors look much much better, but it seems from the last chart for itemized deduction, interest paid in CPS is gone in this version? @andersonfrailey

**Comment by andersonfrailey** [comment]
@Amy-Xu that's because I'm working on slightly tweaking the final prep scripts and left `HMIE`, which I originally thought was equivalent to interest paid, out. `HMIE` is just home mortgage interest expense, part of e19200, but not all of it. It's still much higher than e19200 in the PUF in sum though.

I'm reading through the IRS documentation and there are a couple of instances where HMIE is not fully deductible, so I'm working that into final prep to see the affects on the aggregate total.

**Comment by Amy-Xu** [comment]
@andersonfrailey Sounds great. Thanks!

**Comment by andersonfrailey** [comment]
Update on above comment. All of the instances where home mortgage interest isn't deductible are related to when you took out the mortgage and total home equity so it doesn't look like I'll be able to adjust `HMIE` based on that. Right now it's a little over $100 billion higher than e19200 in the PUF.

**Comment by Amy-Xu** [comment]
@andersonfrailey along the line, I was wondering if HMIE wasn't imputed for the purpose of itemized deduction, it might include interest of standard deductors although this item is not deductible for them. So the total of HMIE might not be reflective how much is taken account in actual itemize deduction. Do you see what I mean? Have you looked at how much in total under the records who take itemized deduction?

**Comment by andersonfrailey** [comment]
@Amy-Xu 
>Have you looked at how much in total under the records who take itemized deduction?
Yep. Both numbers from above are just for itemizers.

**Comment by andersonfrailey** [comment]
Part of the reason refundable credits are so much higher in the CPS than the PUF is the number of children claimed for the EIC. EIC is topped at 3 in the PUF file and I did the same with the CPS, but the almost every tax unit in the CPS file claims 3 children while 90% of the tax units in the PUF claim none at all. Here's a table with the break down of the weighted sums:

|# Children Claimed|	% - CPS|	% - PUF	|CPS |PUF|
|---|---------|--------|----------|--------|
|0	|0.0|0.9|	14,396.5|140,952,059.1|
|1|0.0|0.1|22,451.6|10,061,670.4|
|2|0.0|0.0|162,606.9|7,931,481.3|
|3|1.0|0.0|162,996,315.4|3,279,408.9|

@Amy-Xu getting this fixed will have a huge positive affect in terms of closing the gap in income tax liability. I'll update when it's fixed.

**Comment by Amy-Xu** [comment]
👍 good catch! @andersonfrailey

**Comment by andersonfrailey** [comment]
Issue with the EIC at least partially fixed. The piece of code that counted the number of children that could be claimed for the EIC was also giving children to tax units without any. The [notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb) has been updated to reflect the changes.

The error in income tax liabilities has dropped from -30.7% to -18.7% and the error in total refundable credits has dropped from 153.2% to -21.3%. The CPS is missing a touch over $20 billion in refundable credits right now. Around $8 billion of that is from the American Opportunities Credit because the CPS lacks that data. About $5 billion is the refundable portion of the Child Tax Credit. About $7 billion is due to the EITC now being lower in the CPS than PUF.

**Comment by codykallen** [comment]
@andersonfrailey, these are good improvements in dealing with the EITC problems in the CPS. As you mentioned,

> About $7 billion is due to the EITC now being lower in the CPS than PUF.

If we are imputing the EITC instead of using the error-prone EITC data in the CPS, we should actually expect the CPS EITC amount to be somewhat lower than the PUF EITC amount in aggregate, due to the high error rate in the EITC. Hence imputed EITC payments may be less than actual EITC payments. Officially, the EITC error rate in 2013 was 24% ($14.5 billion), but the actual amount of overpayment is somewhat lower than that, according to the National Taxpayer Advocate.

**Comment by andersonfrailey** [comment]
@codykallen, thanks for the info! Glad we have an explanation for at least some of the difference.

**Comment by Amy-Xu** [comment]
@andersonfrailey this is great progress! I roughly remember John said he wouldn't be too surprised if there's a 20% difference. Personally I feel a bit concerned about itemized deduction. Since Anderson has already gotten that in his plate, @MattHJensen do you think any other cell in diagnostic table need significant adjustment?

**Comment by MattHJensen** [comment]
@Amy-Xu, I discussed a few next steps with Anderson. The primary thing is to add an 'SOI tables' column to the comparison table so that we aren't just comparing against the extrapolated puf, which already has its own error. Other than that, I think this is great, and we should focus on releasing and documenting the file and all steps to creating it now, _and then_, turn back to improving the results.

**Comment by andersonfrailey** [comment]
[Here](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb) is the notebook with SOI comparisons added in.

@Amy-Xu @MattHJensen

**Comment by martinholmer** [comment]
@MattHJensen said in taxdata issue #90 with respect to the new CPS input data for Tax-Calculator:
> I discussed a few next steps with Anderson.  The primary thing is to add an 'SOI tables' column to the comparison table so that we aren't just comparing against the extrapolated puf, which already has its own error.  Other than that, I think this is great, and **we should focus on releasing and documenting the [CPS] file and all steps to creating it now**, and then, turn back to improving the results.

Now that @andersonfrailey has added the SOI tables information to [his CPS notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb), I want to ask him a few questions about next steps:

(1)  Is the directory structure of the taxdata repo that we implemented many months ago going to be sufficient for the CPS data and processing scripts?

(2) How big is the `cps.csv` file?

(3) Do you envision making a preliminary version of it available for review?  

(4) Have you had a chance to generate the `cps_weights.csv` or `cps_ratios.csv` files?

I'm asking (2)-(4) because there may be some design issues that need to be solved so that Tax-Calculator can be made to handle easily either CPS or PUF input data.  Knowing more about the size and contents of the CPS file will help in thinking about those design issues.

@MattHJensen @andersonfrailey @Amy-Xu @feenberg @hdoupe

**Comment by andersonfrailey** [comment]
@martinholmer 

> (1) Is the directory structure of the taxdata repo that we implemented many months ago going to be sufficient for the CPS data and processing scripts?

The current file structure is good for now, but I may make a few changes after I address some of the details I'll talk about in response to your other questions. The scripts John wrote to develop this file and that I've been working on are written in SAS, but they could all be placed in the `cps_data` directory in the same manor we have the matching files in `puf_data`. There are no current plans to translate to python like we did the matching programs.

>(2) How big is the cps.csv file?

Without any of the benefits imputed from C-TAM added on, it is 110.1 MB. With those benefits, it is 167.9 MB. It's significantly larger than the PUF because it has 456,465 records as compared to 215,525 in the PUF.

 >(3) Do you envision making a preliminary version of it available for review?

I'm happy to send the most recent file and it's associated weights to you or anybody else who would like to take a look. If there's interest I could also post a link to a Dropbox folder containing both.

>(4) Have you had a chance to generate the cps_weights.csv or cps_ratios.csv files?

John created the `cps_weights.csv` file. I haven't recreated it myself yet, but he says he used the same process as was used for the CPS matched PUF file, so it's possible the scripts used for `puf_stage2` will also work with the CPS, with some minor tweaks.

As for `cps_ratios.csv`, the AGI variable reported in the CPS is not a reliable measure to perform stage 3 like `e00100` is in the PUF. The highest reported AGI in the CPS is just over $2 million. This leaves the top two AGI bins used in stage 3 completely empty. I have been playing with adding a function to the final prep file that performs an adjustment based on a total income variable created during the matching process. This adjustment is just for one year (2014) and then I set the `adjust_ratios` in Tax-Calculator to `None`. I also adjusted a lot more variables: interest income, qualified and ordinary dividends, and business income.

The resulting distributions aren't bad and in some cases actually a little better than the PUF, as can be seen in the notebook. I haven't tried creating a full `cps_ratios` file to see how it compares for future years yet.

Connecting the above back to question (1). If our current stage 2 scripts work on this CPS, we could change `puf_stage2` to simply `stage2` like we did with stage 1 and use it to create the both weights. Though we would need to modify the script so that it named the file differently depending on which file it is creating weights for. If changes are needs to the stage 2 script, then keeping it in `cps_stage2` would be fine.

For stage 3, because more variables need adjusting, adding modified stage 3 to `cps_stage3` is the ideal solution, assuming we do perform a stage 3 rather than the simple one year adjustment I have been experimenting with.

In sum, we should keep the current structure for now, but it's possible a couple of adjustments will need to be made down the road.

**Comment by martinholmer** [comment]
@andersonfrailey,  Thanks for all the useful information and answers to my questions.  And thanks again for all your work on the CPS file-prep project.

A couple of things:

(a) Even if puf and cps stage 2 logic is similar (or even the same) it would maintain the separate directories: `puf_stage2` and `cps_stage2`.  You never can tell if the script will need to be significantly different in the future.

(b) I would like to get copies of the larger (with benefits) `cps.csv` file and the `cps_weights.csv` file in order to consider how Tax-Calculator would best handle them and to consider storage options (give their size).  I'm not going to do any serious analysis work with them; just experiment with different ways of handling them.  Probably best to put the files in a single zip file to hold down the size and make it easy for you to email them to me.

**Comment by martinholmer** [comment]
@andersonfrailey said:
> Connecting the above back to question (1). If our current stage 2 scripts work on this CPS, we could change `puf_stage2` to simply `stage2` like we did with stage 1 and use it to create the both weights. Though we would need to modify the script so that it named the file differently depending on which file it is creating weights for. If changes are needs to the stage 2 script, then keeping it in `cps_stage2` would be fine.

After more thought, I think this would be a good option if it turns out the stage 2 script is the same for both puf and cps.  If so, I guess you would have to revise the script to handle either puf or cps, including, I would assume, different LP tolerances for puf and cps.  And maybe other differences as well.  But collapsing `puf_stage2` and `cps_stage2` into `stage2` would reduce code duplication which is always a good thing.  So, yes, do keep this option in mind as your work continues.

**Comment by martinholmer** [comment]
@andersonfrailey, Question about the `cps.csv` file contents after info on that file, which is:
```
$ ls -l cps.csv
-rw-r--r--  1 mrh  staff  172826548 Jul  7 08:42 cps.csv

$ awk -F, 'NR>1{n++}NR==1{print NF}END{print n}' cps.csv
84
456465

$ ../csvvars cps.csv
1 age_head
2 age_spouse
3 e00200p
4 e00900p
5 e02100p
6 e00200s
7 e00900s
8 e02100s
9 e00600
10 e00800
11 e01500
12 e02300
13 e02400
14 s006
15 DSI
16 blind_head
17 blind_spouse
18 nu18
19 n1821
20 n21
21 SSI_VAL1
22 SSI_VAL2
23 SSI_VAL3
24 SSI_VAL4
25 SSI_VAL5
26 SSI_VAL6
27 SSI_VAL7
28 SSI_VAL8
29 SSI_VAL9
30 SSI_VAL10
31 SSI_VAL11
32 SSI_VAL12
33 SSI_VAL13
34 SSI_VAL14
35 SSI_VAL15
36 SSI_PROB1
37 SSI_PROB2
38 SSI_PROB3
39 SSI_PROB4
40 SSI_PROB5
41 SSI_PROB6
42 SSI_PROB7
43 SSI_PROB8
44 SSI_PROB9
45 SSI_PROB10
46 SSI_PROB11
47 SSI_PROB12
48 SSI_PROB13
49 SSI_PROB14
50 SSI_PROB15
51 SSI
52 XTOT
53 filer
54 FLPDYR
55 MARS
56 e01100
57 e01400
58 e03300
59 e03270
60 e20400
61 e32800
62 e18500
63 e03240
64 e17500
65 RECID
66 e18400
67 e00200
68 e00900
69 e02100
70 e00650
71 e00300
72 e00400
73 e01700
74 e19800
75 e20100
76 e03210
77 e03150
78 nu13
79 nu05
80 n24
81 elderly_dependent
82 f2441
83 EIC
84 agi_bin

$ 
```
My question is about the variables numbered 21-50 and variable 51, SSI.  They all seem to be about SSI benefits, when I thought you were sending me a file that had all the different benefits imputed.  I admit I haven't been following the development in the C-TAM repo, but I'm confused about why there are so many SSI variables and no variables for other benefits.

@Amy-Xu

**Comment by andersonfrailey** [comment]
@martinholmer whoops it seems like I sent you the file before I changed the final prep scripts to keep in the other benefits. I'll resend it.

All of the other SSI variables are there for @Amy-Xu. She's working on ways to extrapolate the benefits for analysis is future years.

**Comment by andersonfrailey** [comment]
[Updated notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb) contains participation rates, total and average benefits for each of the benefit programs imputed by C-TAM. Also includes a little more clarity with regards to itemized deductions and some comments in the code to show what's going on.

@martinholmer @MattHJensen @Amy-Xu

**Comment by martinholmer** [comment]
@andersonfrailey said:
> whoops it seems like I sent you the file before I changed the final prep scripts to keep in the other benefits. I'll resend it.

Thanks.  I've got the new file you sent and it contains all the imputed benefits.

@andersonfrailey also said:
> All of the other SSI_* variables are there for @Amy-Xu. She's working on ways to extrapolate the benefits for analysis in future years.

I see.  Once she figures out the best way to extrapolate benefit amount, will all these SSI_* variables be removed from the CPS file?  Or, is the plan to always include them?

**Comment by andersonfrailey** [comment]
@martinholmer 
>I see. Once she figures out the best way to extrapolate benefit amount, will all these SSI_* variables be removed from the CPS file? Or, is the plan to always include them?

That will be up to @Amy-Xu. As part of integrating this cps.csv file into Tax-Calc, we want a way to extrapolate those variables. Whether we need to keep those variables in the CPS file or go a different route will ultimately depend on what she thinks the best way forward is.

**Comment by Amy-Xu** [comment]
@martinholmer asked:

>I see. Once she figures out the best way to extrapolate benefit amount, will all these SSI_* variables be removed from the CPS file? Or, is the plan to always include them?

I think they should be dropped. Right now the plan is to have a separate file storing all extrapolated benefits, and read into TC the same way as cps weights.

**Comment by andersonfrailey** [comment]
[Updated notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb) shows the result of taking the HMIE variable in the CPS and shirking it for everyone by the ratio of home interest in [table 2.1](https://www.irs.gov/uac/soi-tax-stats-individual-statistical-tables-by-size-of-adjusted-gross-income) of the SOI stats to HMIE.

@Amy-Xu @MattHJensen

**Comment by Amy-Xu** [comment]
@andersonfrailey Looking at cell 20 in your notebook, it seems the adjusted HMIE is quite a bit lower than PUF or SOI number. Did you use the total of HMIE or the total of itemizers' HMIE to create the ratio?

**Comment by andersonfrailey** [comment]
@Amy-Xu I did total HMIE in the CPS to the HMIE portion of the Interest Paid deduction reported by SOI. I'll do HMIE of just itemizers in the CPS and see what happens

**Comment by Amy-Xu** [comment]
I guess this version confuses me in that CPS total itemized deduction is lower than PUF, but somehow CPS data has more itemizers. In a previous version where CPS has higher aggregates on both interest paid and state and local income taxes, it makes sense that CPS would end up with more itemizers than it should be. But in this version, the total itemized deduction in CPS is lower than PUF, but still end up with more itemizers?

**Comment by Amy-Xu** [comment]
@andersonfrailey Thanks!

**Comment by andersonfrailey** [comment]
@Amy-Xu, using `SOI HMIE / HMIE[CPS|Itemizer]` results in the total for interest paid deduction being about 28% too high

**Comment by Amy-Xu** [comment]
@andersonfrailey I see. My whole point is just to see whether there is a way to get interest paid deduction close to SOI, and theoretically number of itemizers will be closer as a result. If the ratio of `SOI HMIE / HMIE[CPS|Itemizer]` doesn't work, which I guess is people identified as itemizers are not itemizers anymore after their deduction get scaled down, do you think there is a way to get interest paid total similar to SOI? An alternative question, if this is not feasible, is what would be the best adjustment to get the sum of (%itemizers diff + %itemized deduction diff + % standard deduction diff + % standard deductors diff) minimized. Does it make sense?

**Comment by andersonfrailey** [comment]
@Amy-Xu, I'm sure there's a way to get HMIE close to SOI interest paid. In theory this would help minimize `%itemizers diff + %itemized deduction diff + % standard deduction diff + % standard dedicators diff` as well. I'll play around with a few ideas and let you know what I can get.

**Comment by andersonfrailey** [comment]
After playing with the interest paid deduction for awhile, I've updated the [notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb) to show the results of using `HMIE * .8` as `e19200`. I used .8 as a scalar because the other ratios discussed above resulted in `e19200` still being too high. These ratios were around .9 so I decided to give .8 a shot and the resulting `e19200` was pretty much spot on. That being said, home mortgage interest is not the only component of the interest paid deduction, though it does comprise roughly 93% of total reported according to SOI stats, so it might be more accurate to have `e19200` be a little lower in the CPS file than the PUF.

Despite `e19200` being more accurate, total number of itemizers is still much higher than the PUF or the SOI stats. I've tried a number of other ratios to adjust `HMIE` in the CPS file, but haven't had any luck in getting that number down. I'm currently working on plotting out the distribution of itemized deductions and the itemizers themselves. @Amy-Xu suggested the issue may be that more people have just enough in itemized deductions to out value the standard deduction.

I'll post more as I work on this.

cc @MattHJensen

**Comment by Amy-Xu** [comment]
I suspect itemizers in CPS outruns PUF is that CPS has more people w/ smaller amount of itemized deduction than PUF, and less people with large amount of itemized deduction. One way to verify is to tabulate itemizer number by itemize deduction amount @andersonfrailey. If this is true, it means we might need to apply different scaler to people w/ different itemize deduction amount. For example, disqualify some itemizers by reduce their deduction to levels below 6k, and augment the others itemize deduction to match the total admin itemized deduction number. Even though this approach might get us a better total liability number at some point, it seems to me this approach will take longer and is susceptible to accusations being too manipulative.

@MattHJensen What do you think the best way to deal with itemized deduction at this point?

**Comment by hdoupe** [comment]
@Amy-Xu said

> I suspect the reason why the number of itemizers in CPS outruns PUF is that CPS has more people w/ smaller amount of itemized deduction than PUF, and less people with large amount of itemized deduction. One way to verify is to tabulate itemizer number by itemize deduction amount @andersonfrailey. If this is true, it means we might need to apply different scalers to people w/ different itemize deduction amount. For example, disqualify some itemizers by reducing their deduction to levels below 6k, and augment the others itemize deduction to match the total admin itemized deduction number

How do the distributions of the components (medical expenses, interest paid, etc.) of the itemized deduction amount in the CPS compare to those of the PUF? To me, it would make more sense to scale those components than to scale the full itemized deduction amount.

**Comment by Amy-Xu** [comment]
@hdoupe Take a look at Cell 22 in the latest notebook Anderson posted in his comment. I believe what Anderson has been doing is just scaling the interest paid section of itemized deduction, instead of the full itemized deduction. If you trace back a few previous comments by Anderson, you can see that scaling back one item will increase the number of total itemizers more than desired level -- at least that's my understanding. The increased itemizer number decreases total number of standard deductors, and total standard deduction, which drag ind income liability further away from where it's supposed to be. Does that make sense?

**Comment by feenberg** [comment]
On Wed, 12 Jul 2017, Henry Doupe wrote:

> 
> @Amy-Xu said
>
>       I suspect the reason why the number of itemizers in CPS outruns
>       PUF is that CPS has more people w/ smaller amount of itemized
>       deduction than PUF, and less people with large amount of
>       itemized deduction. One way to verify is to tabulate itemizer
>       number by itemize deduction amount @andersonfrailey. If this is
>       true, it means we might need to apply different scalers to
>       people w/ different itemize deduction amount. For example,
>       disqualify some itemizers by reducing their deduction to levels
>       below 6k, and augment the others itemize deduction to match the
>       total admin itemized deduction number
> 
> How do the distributions of the components (medical expenses, interest paid,
> etc.) of the itemized deduction amount in the CPS compare to those of the
> PUF? To me, it would make more sense to scale those components than to scale
> the full itemized deduction amount.

It would help to know where the itemized deduction imputation comes from. 
Is it the CEX?

dan

**Comment by hdoupe** [comment]
@Amy-Xu Ah, ok thanks.  That makes sense.  Sorry, I should've looked before I commented.

**Comment by Amy-Xu** [comment]
@andersonfrailey Since we have been applying scaler, the adjustment might be accused of data manipulation already. One brutal force way to solve this problem I think is to apply one scaler getting total itemizers in right place, and then apply another scaler to augment these itemizers' interest paid & state and local to match the total itemize deduction. But it's Matt's call to do this or not.

@hdoupe no worries.

**Comment by Amy-Xu** [comment]
@feenberg Dan, John has this documentation for CPS tax unit online at: http://www.quantria.com/assets/img/TechnicalDocumentationV4-2.pdf

This documentation (on page 14) seems to indicate two biggest itemized deductions -- home mortgage interest and state and local income taxes -- are respectively imputed from Survey of Consumer Finance (Federal Reserve Board) and a proprietary state calculator. Other items are matched from SOI etc. 

Do you think this mortgage interest issue might be rooted in original CPS income distribution? CPS doesn't have enough high income earners so the mortgage interests matched from SCF are tilted toward lower end.

**Comment by andersonfrailey** [comment]
@Amy-Xu it looks like you're right about there being more itemizers, but smaller totals, as can be seen in the plots below.
![item_rate](https://user-images.githubusercontent.com/20684675/28132090-d04f91f4-6709-11e7-8e5b-b55059626c11.png)
![item_mean](https://user-images.githubusercontent.com/20684675/28132091-d26fdd4a-6709-11e7-94aa-67427c06531f.png)

**Comment by MattHJensen** [comment]
@Amy-Xu asked: 

> @MattHJensen What do you think the best way to deal with itemized deduction at this point?

My view is that we should leave everything as is and focus on documenting the file and how to recreate the file from scratch from inital datasets (CPS, SCF, etc.), including all imputations. Only after the file is documented and have scripts to create it from scratch should we focus on tuning things. At that point we'll have a much better idea of whether the imputations themselves should be adjusted or whether the file should receive 'post processing' adjustments like those discussed above.

**Comment by Amy-Xu** [comment]
@MattHJensen said:

> My view is that we should leave everything as is

What do you mean by 'as is'? You asked to include HMIE as e19200 in this [comment](https://github.com/open-source-economics/UBI-examples/issues/9#issuecomment-314145505). Without any scaling adjustment, this HMIE/e19200 will increase the different of individual income liability between SOI and CPS to more than 30%. Do you mean to include HMIE or not?

**Comment by MattHJensen** [comment]
@Amy-Xu asked:

> Do you mean to include HMIE or not?

If others agree with me that home mortgage interest expense is a component "total interest paid", then I think we should include HMIE in e19200. If others don't, then I'd like to discuss that further. 

>What do you mean by 'as is'? 

My point is that we should line up as many of the variables from the CPS file with the puf as we can and then wait on doing further adjustments until we have documented and can reproduce the CPS file.

Otherwise we will run into more conversations like this:

q. Why did you adjust the interest deduction amount in post processing? 
a. Because the number wasn't close to SOI. 
q. Why wasn't the number close to SOI? 
a. I'm not quitre sure, something to do with the imputation. 
q. How does the imputation work? 
a. I'm not quite sure, here's a link to John O'Hare's documentation.

**Comment by codykallen** [comment]
@MattHJensen said:

> If others agree with me that home mortgage interest expense is a component "total interest paid", then I think we should include HMIE in e19200. If others don't, then I'd like to discuss that further.

It is a component of total interest paid, and the most important one. I would love if we could get the other components as well, but I could settle for at least being able to distinguish between HMIE and non-HMIE deductible interest expense.

**Comment by Amy-Xu** [comment]
@andersonfrailey Could you explain again why you thought [HMIE is not equivalent to e19200](https://github.com/open-source-economics/UBI-examples/issues/9#issuecomment-313699845) so everyone can see?

**Comment by andersonfrailey** [comment]
@Amy-Xu I came to that conclusion by talking to John about it. It is a component of e19200 as @codykallen said, but not a direct map like other variables are.

I generally agree with the points @MattHJensen and @Amy-Xu have made about excessive data manipulation. I suppose it would be best to leave it as is and make a note in the documentation of the dataset about it and the effects it may have on the results coming out of tax-calculator.

**Comment by andersonfrailey** [comment]
[Updated notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb?flush_cache=true) just includes some additional information requested by @Amy-Xu on the number of benefits participants.

**Comment by martinholmer** [comment]
@andersonfrailey said:
> [Updated notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/CPS%20Comparisons.ipynb?flush_cache=true) just includes some additional information requested by @Amy-Xu on the number of benefits participants [and average benefits by wages among participants].

Thanks for the additional tabulations.  

I'm puzzled by the distribution of benefits by wage percentile among participants.  For means-tested programs, I would expect to see benefits decline as wages rise.  But what we see for SSI is nothing like that.  And we see just the opposite for Social Security, which has an individual level "earnings test" (that reduces benefits as wages rise).  I do not expect to see such a decline for benefits that are not means-tested, such as Medicare.  What am I missing?   Is there some explanation that makes sense out of what is puzzling me?

@MattHJensen @Amy-Xu

**Comment by Amy-Xu** [comment]
@martinholmer  Thanks for looking into the charts. At this point, it might be hard to explain the trends in participation and benefit charts because these charts are not final version yet -- we might still need to tweak here and there. That being said, I try to explain what looks sensible to me regarding three observations you have respectively for SSI, Social Security and Medicare. Many are just my speculations -- I'm happy to start the discussion and hear more feedback since there's no official tax-unit distribution we could compare with.

- SSI: The average benefit chart with little benefit decline I believe is plotted with participants only. A very small number of high income taxpayers received SSI for themselves or their families over some periods of the year, during which they might get laid off or lose their jobs for other reasons. Thus, they might get the benefit at same level, making the average benefit among participants chart look flat, but if you take a look at the average benefit among entire population, the benefit decline is still quite significant there since the participation rate of high income groups is very low.

- Social Security: Although Social Security benefit is subject to earning test, I think that applies to people under normal retirement age, which I suspect is not the majority of beneficiaries included in the chart. Also, SS benefit declines as earning goes up, but the earnings refer to historical earnings instead of current wage. Thus I feel it's quite hard to tell how it should look like when most beneficiaries are probably have little current wage income. 

- Medicare: This is something I'm not quite sure about, but I suspect higher income beneficiaries might have relied on private insurance more than lower income beneficiaries, as suggested [in this BLS report](https://www.bls.gov/opub/btn/volume-5/household-healthcare-spending-in-2014.htm). Additionally in the raw CPS imputation, as you can see in [C-TAM documentation](https://github.com/open-source-economics/C-TAM/blob/master/C-TAM%20Documentation%20v0.1.2.pdf), the top quintile indeed has a lower average benefits than the other four quintiles. Last but not least, the medicare average benefit chart has most beneficiaries between $8000 and $16000 benefits. The sharp decline may look not so sharp if the y-axis of the chart is from $0 to $16000 instead.

**Comment by andersonfrailey** [comment]
The CPS files and documentation have been merged into master so I am closing this issue.
